File: norintegrate-mcp/src/main/java/com/norintegrate/mcp/resource/ProcedureResource.java
package com.norintegrate.mcp.resource;

import com.norintegrate.common.procedure.ProcedureService;
import java.util.List;
import org.springframework.ai.mcp.annotation.McpResource;
import org.springframework.stereotype.Component;
import tools.jackson.core.JacksonException;
import tools.jackson.databind.ObjectMapper;

@Component
public class ProcedureResource {

  private final ProcedureService procedureService;
  private final ObjectMapper objectMapper;

  public ProcedureResource(ProcedureService procedureService, ObjectMapper objectMapper) {
    this.procedureService = procedureService;
    this.objectMapper = objectMapper;
  }

  @McpResource(
      uri = "norintegrate://procedures",
      name = "procedures",
      title = "Settlement Procedures",
      description =
          "Complete catalog of Norwegian settlement procedures with responsible authorities,"
              + " estimated processing times, and required documents",
      mimeType = "application/json")
  public String getProcedures() throws JacksonException {
    var summaries =
        procedureService.findAll().stream()
            .map(
                p -> {
                  var docs =
                      procedureService.getDocumentRequirements(p.getId()).stream()
                          .map(d -> new DocumentSummary(d.getDocumentName(), d.isMandatory()))
                          .toList();
                  return new ProcedureSummary(
                      p.getId(),
                      p.getTitle(),
                      p.getDescription(),
                      p.getAuthority(),
                      p.getEstimatedDays(),
                      docs);
                })
            .toList();
    return objectMapper.writeValueAsString(summaries);
  }

  public record ProcedureSummary(
      long id,
      String title,
      String description,
      String authority,
      Integer estimatedDays,
      List<DocumentSummary> documents) {}

  public record DocumentSummary(String documentName, boolean mandatory) {}
}

File: norintegrate-mcp/src/main/kotlin/com/norintegrate/mcp/resource/ProcedureResource.kt
package com.norintegrate.mcp.resource

import com.norintegrate.common.procedure.ProcedureService
import org.springframework.ai.mcp.annotation.McpResource
import org.springframework.stereotype.Component
import tools.jackson.databind.ObjectMapper

@Component
class ProcedureResource(
    private val procedureService: ProcedureService,
    private val objectMapper: ObjectMapper,
) {
    @McpResource(
        uri = "norintegrate://procedures",
        name = "procedures",
        title = "Settlement Procedures",
        description =
            "Complete catalog of Norwegian settlement procedures with responsible authorities," +
                " estimated processing times, and required documents",
        mimeType = "application/json",
    )
    fun getProcedures(): String {
        val summaries =
            procedureService.findAll().map { p ->
                val docs =
                    procedureService.getDocumentRequirements(p.id).map { d ->
                        DocumentSummary(d.documentName, d.isMandatory)
                    }
                ProcedureSummary(p.id, p.title, p.description, p.authority, p.estimatedDays, docs)
            }
        return objectMapper.writeValueAsString(summaries)
    }

    @JvmRecord
    data class ProcedureSummary(
        val id: Long,
        val title: String,
        val description: String?,
        val authority: String?,
        val estimatedDays: Int?,
        val documents: List<DocumentSummary>,
    )

    @JvmRecord
    data class DocumentSummary(
        val documentName: String,
        val mandatory: Boolean,
    )
}
